[Goh Eng Hui, Jeremy] iP

jemehgoh opened the pull request.

jemehgoh committed:
Change chatbot name

jemehgoh committed:
Implement greet and exit

jemehgoh committed:
Add command echo and bye command

jemehgoh committed:
Change display format to arrow

jemehgoh committed:
Add ability to store text entered by user

jemehgoh committed:
Add ability to display stored tasks

jemehgoh committed:
Add ability to mark and unmark tasks as done

jemehgoh committed:
Tweak code to follow Java coding standard

jemehgoh committed:
Add toString() method to Task class

Add the toString() method, and modify the "list", "mark", and "unmark" commands in the Nell class to use it

jemehgoh committed:
Add ToDo class and command to add ToDo task

jemehgoh committed:
Add Deadline class and deadline command to Nell

jemehgoh committed:
Add Event class and event command to Nell

jemehgoh committed:
Add UI tests for up to Level-3 increment

jemehgoh committed:
Add UI tests for ToDo class and todo command

jemehgoh committed:
Add UI tests for Event and Deadline tasks and commands

jemehgoh committed:
Extract command methods from main

The code for the execution of each user command is within a single
switch statement in main().

This makes it difficult to read and trace through the code.

Let's extract the code handling the command executions into separate
methods. This improves the code quality, and also ensures that there is
a single level of abstraction in main() and for the command handling
code.

jemehgoh committed:
Extract greet and command receiving methods from main

The code for greeting the user and receiving user input is written in
one block in main(). This makes the code difficult to read and debug.

Let's:
* extract the code for greeting the user to a greetUser() method
* extract the code for receiving user input into a getCommands() method

This allows for code within main and each of the new methods to be at a
single level of abstraction, increasing code quality and readability.

jemehgoh committed:
Refactor code for checking command validity

The code to check if a selected task is valid is present as separate
code blocks in two different methods. This results in unnecessary code
duplication.

Let's refactor the code to a new method, that checks if the input task
is valid. This makes the code easier to read and makes tracing through
it easier.

jemehgoh committed:
Change method access modifiers in Nell (except main()) to private

jemehgoh committed:
Refactor task adding functions into one method

The addition of tasks to the task list is done by multiple methods, one
each for the Task class or its children. This results in code
duplication, and also terms like "task[taskCount - 1]" which are
difficult to understand.

Let's consolidate the adding of tasks to the task list into a single
method. This reduces code duplication and makes the code easier to
debug.

ChuaCleon commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+     * @param taskToAdd The task to be added to the list
+     */
+    private static void listAddTask(Task taskToAdd) {
+        // Stores text in task list
```
redundant comment here, easily understandable 2 lines of code

ChuaCleon commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+     * Executes a bye command
+     *
+     */
+    private static void executeCommandBye() {
```
good naming convention, however I would suggest you to remove "executeCommand" from your function names (ie just leaving the command name as the function name). This way it improves readability of the code

jemehgoh replied:
This makes sense, but for some of these commands (like "todo" and "deadline"), I believe that following your suggestion would make the method name potentially misleading, and also go against naming conventions (method names being verbs). Would it be better for me to change the method names to "execute<command>" instead?

ChuaCleon commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+     *
+     * @param command The command body
+     */
+    private static void executeCommandAddTask(String command) {
```
Instead of including the execute functions in the Nell class, might want to consider creating a separate class to include these functionalities like (CommandManager or something) to delegate command management and improve abstraction in your code this way

jemehgoh replied:
I agree with this, thanks for your suggestion!

ChuaCleon commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+    private static void executeCommandDeadline(String commandBody) {
+        System.out.println("-> The task has been added to the list:");
+        String[] commandWords = commandBody.split("/by");
+        Deadline deadlineToAdd = new Deadline(commandWords[0].trim(), commandWords[1].trim());
```
good and simple implementation, easily understandable code

Yisoong commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+     * list command
+     *
+     */
+    private static void executeCommandList() {
```
I noticed that "executeCommand" is repeated a lot in the naming of your functions and it might be better to shorten the name to focus on what the function is doing.

Yisoong commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+    private static void executeCommandUnmark(String commandBody) {
+        int taskIndex = Integer.parseInt(commandBody);
+        if (checkIfTaskValid(taskIndex)) {
+            // Marks a task as not done
```
This comment is redundant.

Yisoong commented on src/main/java/Nell.java:
```diff
@@ -0,0 +1,217 @@
+     *
+     * @param commandBody The command body
+     */
+    private static void executeCommandEvent(String commandBody) {
```
commandBody sounds a bit clunky and confusing. It might be better to call it "description" for clarity.

jemehgoh replied:
Thanks for your comment, I agree with your suggestion.

Yisoong commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+    }
+
+    public String toString() {
+        return String.format("[E][%s] %s (from: %s to: %s)", this.getStatusIcon(),
```
This line is confusing and seems to be doing a lot in one line. Perhaps it would be better to break it up or use another way to achieve this result.

jemehgoh replied:
I'm personally not sure of how to go about this. Perhaps I could use a taskType variable (in the parent class), and move some of the functionality into the parent class?

Yisoong reviewed (commented):
Good job with your IP! I think some of the naming conventions could be made clearer but overall, great code! Jiayous!!

jemehgoh committed:
Remove unnecessary comments

jemehgoh committed:
Rename command execution methods

The methods for executing the user commands have names beginning with
"executeCommand". This makes the method names difficult to read,
and potentially misleading (as one could think that the methods only
executed the commands)

Let's rename these methods with names that reflect what is actually
done when executing these commands. This way, the method names would be
self-explanatory and more readable.

jemehgoh committed:
Rename user command method parameters

The methods for the user commands share the same parameter
"commandBody". This is vague and does not indicate what the parameter
does in the method.

Let's give the parameters different names, that align with what the
parameter is used for in the methods. This makes for more readable
code, and for the parameter's use to be self-explantory.

jemehgoh committed:
Rename executeCommandBye() to sayBye()

jemehgoh committed:
Add number format error handling to mark and unmark methods

jemehgoh committed:
Add invalid command and incorrect format error handling

jemehgoh committed:
Replace checkIfTaskValid() method with exception

jemehgoh committed:
Add additional error handling for improperly entered commands

jemehgoh committed:
Add error handling for null inputs

jemehgoh committed:
Update test cases to reflect error checking

jemehgoh committed:
Add type member to Task class

jemehgoh committed:
Merge branch 'branch-Level-5'

jemehgoh committed:
Move classes under "nell" package

jemehgoh committed:
Move Task class and its child classes into "nell.task" package

jemehgoh committed:
Add test case for invalid deadline command

jemehgoh committed:
Merge branch 'branch-A-Packages'

jemehgoh committed:
Create new TaskList class

jemehgoh committed:
Refactor task list in Nell to use TaskList class

jemehgoh committed:
Replace ListAddTask() method in Nell with addTask() method in TaskList

The addition of tasks to the task list is done using the ListAddTask()
method, which then calls the addTask() method. This leads to
unnecessary code duplication.

Let's replace the calls to ListAddTask() in Nell with calls to
addTask(), and remove the ListAddTask() method entirely. This makes for
reduced code duplication and more readable code.

jemehgoh committed:
Move message printing from Nell to addTask() method in TaskList

The methods to add various task types in Nell print out the same
messages, and only differ in the types of task they add. This results
in unnecessary code duplication

Let's move the message printing code from these methods to the addTask()
method in TaskList. This allows for reduced code duplication and more
readable code.

jemehgoh committed:
Fix runtest.bat to work with new packages

jemehgoh committed:
Refactor toString() methods in Task class and child classes

jemehgoh committed:
Merge branch 'branch-TaskList'

jemehgoh committed:
Replace tasks array in TaskList with ArrayList

jemehgoh committed:
Refactor task type initialisation to parent class

jemehgoh committed:
Add removeTask() method to taskList

jemehgoh committed:
Remove redundant NullPointerException handling code

jemehgoh committed:
Amend removeTask method in TaskList to print messages on task removal

jemehgoh committed:
Add removeTask method in Nell

jemehgoh committed:
Add code to handle user command to remove tasks

jemehgoh committed:
Amend test cases to account for remove command

jemehgoh committed:
Add writeListToFile method to TaskList

jemehgoh committed:
Add ability to save task list upon chatbot shutdown

jemehgoh committed:
Amend data format in data.txt

The tasks in data.txt are stored with some attributes in square
brackets, and others in parentheses, for readability. This makes
data.txt difficult to parse during initialisation.

Let's store the attributes in lines, separated by pipes.

This makes data.txt easier to parse, allowing it to be read into the
program upon startup.

jemehgoh committed:
Add loadTask method to TaskList

jemehgoh committed:
Add overloaded constructor with isDone parameter with overrides

jemehgoh committed:
Add functionality to load saved files upon startup

jemehgoh committed:
Remove message print for main path in saveToFile method in Nell

jemehgoh committed:
Remove unused getDescription method in Task

jemehgoh committed:
Merge branch 'branch-Level-6'

jemehgoh committed:
Merge branch 'branch-Level-7'

jemehgoh committed:
Update .gitignore to not track JAR

jemehgoh committed:
Update loadFromFile method to handle case where data/data.txt is absent

jemehgoh committed:
Change getFileLine methods to prevent index out of bound

Mahesh1772 commented on src/main/java/nell/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+            return 'X';
+        }
+
+        return ' ';
```
Consider using ternary operator, to for simpler readability.

Mahesh1772 commented on src/main/java/nell/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public String toString() {
+        return String.format("[%S][%s] %s", this.type, this.getStatusIcon(), this.description);
```
Good use of string formatting!

Mahesh1772 commented on src/main/java/nell/tasks/ToDo.java:
```diff
@@ -0,0 +1,17 @@
+    public ToDo(String description, boolean isDone) {
+        super(description, isDone);
+        this.type = "T";
+    }
```
Consider adding Javadocs for these constructors as well.

Mahesh1772 commented on src/main/java/nell/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    protected String type;
```
Consider changing 'type' to a final, if it will not be modified.

Mahesh1772 commented on src/main/java/nell/tasks/Event.java:
```diff
@@ -0,0 +1,28 @@
+    public String getFileLine() {
+        return String.format("%s|%s|%s", super.getFileLine(), this.from, this.to);
+    }
+}
```
Consider adding JavaDocs

Mahesh1772 commented on src/main/java/nell/Nell.java:
```diff
@@ -0,0 +1,346 @@
+            switch (commandWords[0]) {
+            case "bye":
+                sayBye();
+                saveToFile();
```
Try to save tasks at the end of each operation, so that even when the app is not closed properly it retains the changes made.

Mahesh1772 commented on src/main/java/nell/Nell.java:
```diff
@@ -0,0 +1,346 @@
+        String taskType = taskParameters[0];
+        String taskDescription = taskParameters[2];
+
+        boolean taskIsDone = false;
+        if (taskParameters[1].equals("X"))
+        {
+            taskIsDone = true;
+        }
```
Could be simplified to one line: boolean taskIsDone = taskParameters[1].equals("X");

Mahesh1772 reviewed (commented):
The code is clean, follows conventions, and is well-structured. I’ve provided a few suggestions mainly to improve clarity, conciseness, and documentation—nothing critical. 

Consider making the methods in the main file shorter for better readability.

Your attention to detail and consistent coding practices are evident. Keep up the fantastic effort!

jemehgoh committed:
Simplify assignment statement for TaskIsDone variable in Nell

jemehgoh committed:
Simplify method body for getStatusIcon with ternary operator

jemehgoh committed:
Change save to file to be executed after each command

jemehgoh committed:
Remove redundant import statement

jemehgoh committed:
Update gitignore to ignore data file created by saving

jemehgoh committed:
Merge branch 'branch-pr-review-fix'

jemehgoh committed:
Move PrintTaskAtIndex method functionality to TaskList class

jemehgoh committed:
Create Storage class to handle file loading and saving

jemehgoh committed:
Fix javadoc for Storage class constructor

jemehgoh committed:
Move code to load and save task data to Storage class

The methods for loading task data from and saving to a data file are
located in the Nell class. This makes the code in Nell difficult to
read.

Let's create a new Storage class to handle these tasks. This makes the
code in Nell more readable and easier to debug.

jemehgoh committed:
Update runtest.bat to account for new nell.storage package

jemehgoh committed:
Remove unused import statements in Nell

jemehgoh committed:
Add javadoc for Storage class

jemehgoh committed:
Move message constants in Nell to Messages container class

jemehgoh committed:
Convert magic literals in Nell to constants in Messages

jemehgoh committed:
Convert message literals in Storage to constants in Messages

jemehgoh committed:
Convert message literals in TaskList to constants in Messages

jemehgoh committed:
Update runtest.bat to work with new common package

jemehgoh committed:
Add containsKeyword method to Task class

jemehgoh committed:
Add getMatchingTaskList method to TaskList

jemehgoh committed:
Add findMatchingTasks method to Nell

jemehgoh committed:
Add command for user to find tasks using keywords and error message

jemehgoh committed:
Update handleIncorrectInput method to include find command

jemehgoh committed:
Create new Parser class to handle parsing of commands

jemehgoh committed:
Move command parsing and execution to Parser class

The parsing and execution of commands in the Nell program is done by
several methods in Nell. This makes the code in Nell difficult to
read.

Let's move the code for command parsing and execution to another class,
Parser. This makes the code more readable and easier to debug.

jemehgoh committed:
Update runtest.bat to work with new parser package

jemehgoh committed:
Split parseCommand method into multiple methods

The parsing of commands in Parser is done in a single, long
parseCommand method. This makes for a long method that is difficult to
read.

Let's split the parseCommand method into two methods, one for
single-word commands and one for multi-word commands. This makes the
code in the resulting methods and Parser more readable.

jemehgoh committed:
Create new Ui class to handle user interaction in Nell

jemehgoh committed:
Add dataStorage member to Ui class to handle saving of data

jemehgoh committed:
Move handling of user interactions to Ui class

jemehgoh committed:
Create abstract parent Command class

jemehgoh committed:
Create new ListCommand class

jemehgoh committed:
Create new ToDoCommand class

jemehgoh committed:
Add new DeadlineCommand class

jemehgoh committed:
Add javadoc comments and made members in Deadline final

jemehgoh committed:
Create new EventCommand class

jemehgoh committed:
Create new MarkCommand class

jemehgoh committed:
Create new UnmarkCommand class

jemehgoh committed:
Update javadoc for execute method in UnmarkCommand class

jemehgoh committed:
Create new RemoveCommand class

jemehgoh committed:
Create new IncorrectCommand class

jemehgoh committed:
Refactor command execution to use methods in Command child classes

The code for command execution comprises multiple methods in Parser.
This makes the code in Parser difficult to read, and it does not make
sense for command execution code to be within a Parser class.

Let's move the command execution code to separate Command classes, one
for each command. This makes the code more organised and easier to
read.

jemehgoh committed:
Replace wildcard import statement with single class imports

jemehgoh committed:
Remove unused command execution methods in Parser

jemehgoh committed:
Remove unused import statements in Nell

jemehgoh committed:
Remove unused import statements in Parser

jemehgoh committed:
Add additional constructor for IncorrectCommand with message parameter

jemehgoh committed:
Update exception handling code in Parser to use IncorrectCommand class

jemehgoh committed:
Add new parseCommandWithIndex method to Parser

jemehgoh committed:
Rewrote javadoc of command parsing methods in Parser

jemehgoh committed:
Fix javadoc of parseCommandWithIndex method in Parser

jemehgoh committed:
Replace switch statement in parseSingleWordCommand method with if-else

jemehgoh committed:
Extract magic numbers in parseCommand in Parser

jemehgoh committed:
Make constants in Parser static

jemehgoh committed:
Update runtest.bat to account for ui and command packages

jemehgoh committed:
Add .trim method calls to DeadlineCommand constructor

jemehgoh committed:
Remove blank line from test cases

jemehgoh committed:
Add .trim() method to ToDoCommand constructor

jemehgoh committed:
Make unchanged members in Parser and Ui final

jemehgoh committed:
Make unchanged members in Deadline and Event final

jemehgoh committed:
Merge branch 'branch-A-MoreOOP'

jemehgoh committed:
Made 'by' member in Deadline class final

jemehgoh committed:
Merge branch 'master' into branch-Level-8

jemehgoh committed:
Add DateTimeFormatter objects for formatting dates and times for I/O

jemehgoh committed:
Add date and time handling for by member in Deadline

jemehgoh committed:
Update error message for date errors

jemehgoh committed:
Add date and time handling for the Event class

jemehgoh committed:
Merge branch 'master' into branch-Level-9

jemehgoh committed:
Add javadoc to methods in FindCommand

jemehgoh committed:
Replace wildcard import with single class imports

jemehgoh committed:
Add import statement for FindCommand in Parser

jemehgoh committed:
Amend invalid command message to include find command

jemehgoh committed:
Make Task class abstract and create new abstract isOnDate method

jemehgoh committed:
Add implementation for isOnDate to ToDo

jemehgoh committed:
Implement isOnDate method in Deadline

jemehgoh committed:
Implement isOnDate method in Event

jemehgoh committed:
Add getTasksOnDate method to TaskList

jemehgoh committed:
Add new SearchCommand class, associated date format and message

jemehgoh committed:
Add code to handle search command parsing to Parser

jemehgoh committed:
Move exception handling for DateTimeParseException to parseCommand

The exception handling for DateTimeParseException is done separately
for each command in parseMultiWordCommand, but executes the same
code block. This results in duplicate code.

Let's move the exception handling to parseCommand, handling all cases
at once. This reduces code duplication.

jemehgoh committed:
Add javadoc for Task class

jemehgoh committed:
Merge branch 'master' into branch-A-JavaDoc

jemehgoh committed:
Merge pull request #1 from jemehgoh/branch-Level-8

Level-8

jemehgoh committed:
Merge branch 'master' into branch-Level-9

jemehgoh committed:
Update incomplete javadoc header for isOnDate in Deadline

jemehgoh committed:
Merge branch 'master' into branch-Level-9

jemehgoh committed:
Move TaskList into new nell.list package

jemehgoh committed:
Replace commandWord member in Command with COMMAND_WORD constants

jemehgoh committed:
Extract magic "bye" string in Ui as constant

jemehgoh committed:
Rename exception variables

jemehgoh committed:
Amend getMatchingTaskList method to use getTaskStringAtIndex method

jemehgoh committed:
Extract task type specifiers as constants in Task child classes

jemehgoh committed:
Extract magic string message in FindCommand

jemehgoh committed:
Refactor import statement to single class import

jemehgoh committed:
Fix constructor in IncorrectCommand

jemehgoh committed:
Amend class definition for Deadline, Event and ToDo

jemehgoh committed:
Update runtest.bat to account for list package

jemehgoh committed:
Merge pull request #2 from jemehgoh/branch-Level-9

Level-9

jemehgoh committed:
Merge branch 'master' into branch-A-JavaDoc

jemehgoh committed:
Add Override annotation to execute method definitions

jemehgoh committed:
Add javadoc for methods in the ToDo class

jemehgoh committed:
Remove redundant toString method in ToDo

jemehgoh committed:
Add javadoc for methods in Deadline

jemehgoh committed:
Add javadoc for methods in Event

jemehgoh committed:
Add javadoc to methods in TaskList

jemehgoh committed:
Add javadoc to methods in Task

jemehgoh committed:
Amend header comment for getFileLine to bring in line with guidelines

jemehgoh committed:
Fix header comments in Ui

jemehgoh committed:
Remove redundant comments in getCommands method in Ui

jemehgoh committed:
Merge pull request #3 from jemehgoh/branch-A-JavaDoc

A-JavaDoc

jemehgoh committed:
Change error message for search to reflect correct date format

jemehgoh committed:
Update FindCommand to only take in the first keyword entered

jemehgoh committed:
Amend containsKeyword method in Task to only compare entire words

jemehgoh committed:
Add user guide

jemehgoh committed:
Add "Notes on commands" section to user guide

jemehgoh committed:
Change exit command in user guide to reflect actual functionality

jemehgoh committed:
Add line breaks and fix indentation in user guide

jemehgoh committed:
Add whitespace and output for list and search commands

jemehgoh committed:
Add contents to user guide

jemehgoh committed:
Add link to download jar

jemehgoh committed:
Update ToDO command word in user guide to reflect actual functionality

jemehgoh committed:
Move error handling code to parseSingleWordCommand method

The error handling for multi-word commands is done in
parseMultiWordCommands. This leads to the wrong error message being
printed for some multi-word commands.

Let's move the error handling code to parseSingleWordCommand. This
ensures that the correct error messages are printed every time.

jemehgoh committed:
Update FindCommand and SearchCommand to not print additional newline

jemehgoh committed:
Update test cases to account for new functionality

jemehgoh committed:
Update user guide to note case-sensitivity of command words

jemehgoh committed:
Fix term used in user guide

jemehgoh committed:
Update docs/README.md to contain user guide

jemehgoh committed:
Make non-constant variables not final

jemehgoh committed:
Make methods in Nell class not static

jemehgoh committed:
Update Nell to follow coding standard

jemehgoh committed:
Make instance members final

jemehgoh committed:
Update user guide regarding functionality of single-word commands
